src/raftstore/store/snap/builder.rs

```diff
@@ -0,0 +1,558 @@
+        start_key: &[u8],
+        end_key: &[u8],
+        io_limiter: Option<&IOLimiter>,
+        snap_key: SnapKey,
+        snap_stale_checker: &SnapStaleChecker,
+    ) -> Result<usize> {
+        let (mut cf_key_count, mut cf_size) = (0, 0);
+        let mut stale = stale_for_generate(snap_key, snap_stale_checker);
+        match self.tmp_cf_file {
+            Either::Left(ref mut file) => {
+                box_try!(
+                    db_snap.scan_cf(self.cf, start_key, end_key, false, |key, value| {
+                        cf_key_count += 1;
+                        cf_size += key.len() + value.len();
+                        file.encode_compact_bytes(key)?;
+                        file.encode_compact_bytes(value)?;
+                        stale = stale_for_generate(snap_key, snap_stale_checker);
+                        Ok(!stale)
+                    })
+                );
+                // use an empty byte array to indicate that cf reaches an end.
+                file.encode_compact_bytes(b"")?;
+                file.flush()?;
+            }
+            Either::Right(ref mut writer) => {
+                let mut bytes = 0;
+                let base = io_limiter.map_or(0, |l| l.get_max_bytes_per_time());
+                box_try!(
+                    db_snap.scan_cf(self.cf, &start_key, &end_key, false, |key, value| {
+                        let l = key.len() + value.len();
+                        cf_key_count += 1;
+                        cf_size += l;
+                        if let Some(ref limiter) = io_limiter {
+                            if bytes >= base {
+                                bytes = 0;
+                                limiter.request(base);
+                            }
+                            bytes += l as i64;
+                        }
+                        writer.put(key, value)?;
+                        stale = stale_for_generate(snap_key, snap_stale_checker);
+                        Ok(!stale)
+                    })
+                );
+                if cf_key_count > 0 {
+                    box_try!(writer.finish());
+                }
+            }
+        }
+
+        if stale {
+            error!("{} build_snapshot meets stale db snap", snap_key);
+            return Err(Error::Stale);
+        }
+
+        // TODO: calculate written bytes and crc32 while scaning.
+        self.written_bytes = get_file_size(&self.tmp_cf_path)?;
+        self.digest = Digest::new_with_initial(crc32::IEEE, calc_crc32(&self.tmp_cf_path)?);
+
+        SNAPSHOT_CF_KV_COUNT
+            .with_label_values(&[self.cf])
+            .observe(cf_key_count as f64);
+        SNAPSHOT_CF_SIZE
+            .with_label_values(&[self.cf])
+            .observe(cf_size as f64);
+        info!(
+            "[region {}] scan snapshot on cf {}, key count {}, size {}",
+            region_id, self.cf, cf_key_count, cf_size
+        );
+        Ok(cf_key_count)
+    }
+}
+
+struct SnapshotBase {
```

Comment:
it is weird that there are two SnapshotBase structures, although they are in different files, they still belong to the snapshot module.

Reply:
One for read and one for write. Maybe rename them to `SnapshotReadBase` and `SnapshotWriteBase` is better?